### PR TITLE
Incremental builds

### PR DESCRIPTION
**Pull request by stefanloerwald**
How it works:

By defining inputs/outputs for static assets, we can enable incremental builds (note: with VS preview, I'm currently not getting the correct behavior, but `dotnet build` in CLI works as expected. I'm very much suspecting a bug in VS)
```xml
<ItemGroup>
    <InputStaticAssetsCss Include="**\*.css;**\*.scss" Exclude="wwwroot\**\*;obj\**\*;bin\**\*" />
    <OutputStaticAssetsCss Include="wwwroot\bundled.min.css;wwwroot\unbundled.min.css;wwwroot\unbundled.css;" />
</ItemGroup>
```

We run npm install using docker, which makes sure that if `package.json` didn't change, there won't be the need to re-execute that command. Result gets saved as a named docker image `mb_static_assets`.
```xml
<Target Name="BuildNpmBase">
    <!-- Build the base for both styles and scripts: this is essentially just the npm install -->
    <Exec Command="docker build -q -t mb_static_assets -f npm.df ." />
</Target>
```

The actual build of css (and js the same), works with docker too. It requires a few steps

- We copy the results to a tmp dir in `obj`, so we first delete all files in that tmp dir
- We build the static assets with docker, so we also get caching goodness here too. The result gets saved as a docker image named `mb_static_assets_css`
- To copy the files out, we need to create a container from that image
- We then copy out all items from the `wwwroot` folder into the tmp dir in `obj`
- We can now remove the docker container again (that's also done before the container creation while ignoring errors, in case a previous build failed and left the container).
- To make sure we don't get wrong timestamps, we quickly touch those assets
- Then they're copied to the `wwwroot` folder.

```xml
<Target Name="BuildStyles" BeforeTargets="BeforeBuild" Inputs="@(InputStaticAssetsCss)" Outputs="@(OutputStaticAssetsCss)" DependsOnTargets="BuildNpmBase">
    <!-- We need to remove our tmp output dir, otherwise the copy command from docker to host will mess things up -->
    <Message Importance="high" Text="$(BaseIntermediateOutputPath)static_assets_css" />
    <ItemGroup>
        <TmpFiles Include="$(BaseIntermediateOutputPath)static_assets_css\*" />
    </ItemGroup>
    <Delete Files="@(TmpFiles)" />
    <RemoveDir Directories="$(BaseIntermediateOutputPath)static_assets_css" />
    <!-- This is the actual build of assets -->
    <Exec Command="docker build -q -t mb_static_assets_css -f npm_css.df ." />
    <!-- Now we want to copy all output assets from the container to the host -->
    <!-- Remove container instance in case anything went wrong in a previous run, so nevermind if it fails -->
    <Exec Command="docker rm -f tmp_css" ContinueOnError="true" />
    <!-- Create a fresh container instance -->
    <Exec Command="docker create -ti --name tmp_css mb_static_assets_css bash" />
    <!-- Copy relevant output files into tmp folder -->
    <Exec Command="docker cp tmp_css:/src/wwwroot &quot;$(ProjectDir)$(BaseIntermediateOutputPath)static_assets_css&quot;" />
    <!-- Remove container instance -->
    <Exec Command="docker rm -f tmp_css" />
    <ItemGroup>
        <GeneratedStaticAssetsCss Include="$(ProjectDir)$(BaseIntermediateOutputPath)static_assets_css\*" />
    </ItemGroup>
    <!-- Copy files from tmp folder to wwwroot -->
    <Touch Files="@(GeneratedStaticAssetsCss)" ForceTouch="true" />
    <Copy SourceFiles="@(GeneratedStaticAssetsCss)" DestinationFolder="$(ProjectDir)wwwroot" />
    <RemoveDir Directories="$(ProjectDir)$(BaseIntermediateOutputPath)static_assets_css" />
</Target>
```

Note: `Material.Blazor/webpack.config.js` had a typo in it: `scripts` vs. `Scripts`.

### CODE REVIEW

File: Material.Blazor/Material.Blazor.csproj [hunk 1]
```diff
@@ -1,108 +1,117 @@
...
-    <RepositoryUrl>https://github.com/Material-Blazor/Material.Blazor</RepositoryUrl>
-    <PackageLicenseExpression>MIT</PackageLicenseExpression>
-    <PackageReleaseNotes></PackageReleaseNotes>
-    <NeutralLanguage>en</NeutralLanguage>
-    <Copyright>Dioptra and Optimium Health</Copyright>
-    <PackageProjectUrl>https://material-blazor.com</PackageProjectUrl>
-    <Product>Material.Blazor</Product>
-    <UserSecretsId>d7509b6c-bc5c-4a56-a750-09c8ae4aeb85</UserSecretsId>
-  </PropertyGroup>
+	<PropertyGroup>
```
Comment (stefanloerwald) [inline Material.Blazor/Material.Blazor.csproj:3]: Apologies for formatting the entire document. Actual changes only further down...

File: Material.Blazor/Material.Blazor.csproj [hunk 2]
```diff
@@ -1,108 +1,117 @@
...
-    <Exec Command="npm run build-scripts" />
+	<ItemGroup>
+		<None Update="Scripts\MaterialBlazor.ts">
+			<SubType>Code</SubType>
+		</None>
+	</ItemGroup>
 
-    <Exec Command="echo ***** npm run build-js" />
-    <Exec Command="npm run build-js" />
+	<Target Name="BuildNpmBase">
```
Comment (stefanloerwald) [inline Material.Blazor/Material.Blazor.csproj:42]: Here is the beginning of actual changes

File: Material.Blazor/npm.df.dockerignore [hunk 3]
```diff
@@ -0,0 +1,10 @@
+**/*.csproj
```
Comment (stefanloerwald) [inline Material.Blazor/npm.df.dockerignore:1]: The dockerignore files are important to not get any "changed environment -> rebuild" effects using docker

File: Material.Blazor/webpack.config.js [hunk 4]
```diff
@@ -5,7 +5,7 @@ module.exports = {
 
     entry: {
         'MaterialBlazor': [
-            './scripts/MaterialBlazor.ts'
+            './Scripts/MaterialBlazor.ts'
```
Comment (stefanloerwald) [inline Material.Blazor/webpack.config.js:8]: This caused an issue in the linux environment

**Comment by MarkStega** [comment]
Stefan,
The work you did on this was really useful. We incorporated the checks for inputs/outputs of static assets (as well as the removal of fastupdatechecks) into Blazor. After discussion with Simon our conclusion was that the container approach didn't give us enough added benefit & we stuck with using npm directly.

Thanks for the work, it is greatly appreciated.

**Pull request closed without merging**